### PR TITLE
Add Tensorflow Swin model

### PR DESCRIPTION
**Pull request by amyeroberts**
# What does this PR do?

Adds a tensorflow implementation of the Swin architecture and associated tests. 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Will tag specific members / contributors when moved from drafts.

### CODE REVIEW

File: tests/models/swin/test_modeling_tf_swin.py [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+            AutoFeatureExtractor.from_pretrained("microsoft/swin-tiny-patch4-window7-224")
+            if is_vision_available()
+            else None
+        )
+
+    @slow
+    def test_inference_image_classification_head(self):
+        model = TFSwinForImageClassification.from_pretrained(
+            "microsoft/swin-tiny-patch4-window7-224",
+            # FIXME - remove once model has been reviewed and weights pushed to the hub
```
Comment (amyeroberts) [inline tests/models/swin/test_modeling_tf_swin.py:348]: Adding comment to make sure this is removed before merging in

File: tests/models/swin/test_modeling_tf_swin.py [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+            from_pt=True,
+        )
```
Comment (amyeroberts) [inline tests/models/swin/test_modeling_tf_swin.py:350]: Suggestion be be added once model architecture approved and weights pushed to the hub

```suggestion
        model = TFSwinForImageClassification.from_pretrained(
            "microsoft/swin-tiny-patch4-window7-224",
        )
```

**Comment by amyeroberts** [comment]
@LysandreJik @Rocketknight1 @NielsRogge I added you as reviewers to cover the transformers, vision and tensorflow aspects of this PR. Apologies if you're not the right person to review this - and please feel free to remove yourselves or add others :)

**Comment by Rocketknight1** [comment]
I think those are good choices! Reviewing now.

File: src/transformers/models/swin/__init__.py [hunk 3]
```diff
@@ -40,6 +40,14 @@
...
     ]
 
+if is_tf_available():
+    _import_structure["modeling_tf_swin"] = [
+        "TF_SWIN_PRETRAINED_MODEL_ARCHIVE_LIST",
+        "TFSwinForImageClassification",
+        "TFSwinForMaskedImageModeling",
+        "TFSwinModel",
+        "TFSwinPreTrainedModel",
+    ]
```
Comment (NielsRogge) [inline src/transformers/models/swin/__init__.py:50]: This will need to be in a try - except block, similar to the one defined above.

File: src/transformers/models/swin/__init__.py [hunk 4]
```diff
@@ -58,6 +66,15 @@
...
 
+    if is_tf_available():
+        from .modeling_tf_swin import (
+            TF_SWIN_PRETRAINED_MODEL_ARCHIVE_LIST,
+            TFSwinForImageClassification,
+            TFSwinForMaskedImageModeling,
+            TFSwinModel,
+            TFSwinPreTrainedModel,
+        )
+
```
Comment (NielsRogge) [inline src/transformers/models/swin/__init__.py:77]: Same comment here.

**Review by Rocketknight1 — APPROVED** [review]
Overall this looks very good to me, and the tests look quite complete! Can we add a test which will require padding, to ensure that the `maybe_pad` code paths get checked?

File: src/transformers/models/swin/modeling_tf_swin.py [hunk 5]
```diff
@@ -0,0 +1,1418 @@
...
+        pad_right = (self.window_size - width % self.window_size) % self.window_size
+        pad_bottom = (self.window_size - height % self.window_size) % self.window_size
+        pad_values = tf.constant(
+            [
+                [0, 0],
+                [0, pad_bottom],
+                [0, pad_right],
+                [0, 0],
+            ]
+        )
```
Comment (sgugger) [inline src/transformers/models/swin/modeling_tf_swin.py:715]:
```suggestion
        pad_values = tf.constant([[0, 0], [0, pad_bottom], [0, pad_right], [0, 0]])
```
Fits in one line.

File: tests/models/swin/test_modeling_tf_swin.py [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+        result = model(pixel_values, labels=labels)
+        self.parent.assertEqual(result.logits.shape, (self.batch_size, self.type_sequence_label_size))
+
+    def prepare_config_and_inputs_for_common(self):
+        config_and_inputs = self.prepare_config_and_inputs()
+        (
+            config,
+            pixel_values,
+            labels,
+        ) = config_and_inputs
```
Comment (sgugger) [inline tests/models/swin/test_modeling_tf_swin.py:156]:
```suggestion
        config, pixel_values, labels = config_and_inputs
```

File: tests/models/swin/test_modeling_tf_swin.py (hunk 2 above)
Comment (sgugger) [reply tests/models/swin/test_modeling_tf_swin.py:350]: Would suggest one line since it fits ;-)
```suggestion
        model = TFSwinForImageClassification.from_pretrained("microsoft/swin-tiny-patch4-window7-224")
```

**Review by sgugger — APPROVED** [review]
Very clean! Just have a few styling comments on top of Niels' comments for the init.
Thanks a lot!

**Review by Rocketknight1 — APPROVED** [review]
New tests look good to me, so I'm happy to approve this one!
